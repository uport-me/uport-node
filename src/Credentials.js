import { createJWT, verifyJWT, SimpleSigner } from 'did-jwt'
import UportLite from 'uport-lite'
const MNID = require('mnid')
import UportDIDResolver from 'uport-did-resolver'
import MuportDIDResolver from 'muport-did-resolver'
import EthrDIDResolver from 'ethr-did-resolver'
import { toEthereumAddress } from 'did-jwt/lib/Digest'
import { ec as EC } from 'elliptic'
const secp256k1 = new EC('secp256k1')
/**
*    The Credentials class allows you to easily create the signed payloads used in uPort inlcuding
*    credentials and signed mobile app requests (ex. selective disclosure requests
*    for private data). It also provides signature verification over signed payloads and
*    allows you to send push notifications to users.
*/
class Credentials {

  /**
   * Instantiates a new uPort Credentials object
   *
   * The following example is just for testing purposes. You should never store a private key in source code.
   *
   * @example
   * import { Credentials } from 'uport'
   * const credentials = new Credentials({
   *   privateKey: '74894f8853f90e6e3d6dfdd343eb0eb70cca06e552ed8af80adadcc573b35da3'
   * })
   *
   * The above example derives the public key used to generate the did, so only a private key is needed.
   * Generating a public key from a private key is slow. It is recommended to configure the `did` option as well.
   *
   * @example
   * import { Credentials } from 'uport'
   * const credentials = new Credentials({
   *   did: 'did:ethr:0xbc3ae59bc76f894822622cdef7a2018dbe353840',
   *   privateKey: '74894f8853f90e6e3d6dfdd343eb0eb70cca06e552ed8af80adadcc573b35da3'
   * })
   *
   * It is recommended to store the address and private key in environment variables for your server application
   *
   * @example
   * import { Credentials, SimpleSigner } from 'uport'
   * const credentials = new Credentials({
   *   did: process.env.APPLICATION_DID,
   *   signer: SimpleSigner(process.env.PRIVATE_KEY)
   * })
   *
   * Instead of a private key you can pass in a [Signer Functions](https://github.com/uport-project/did-jwt#signer-functions) to
   * present UX or call a HSM.

   * @example
   * import { Credentials } from 'uport'
   *
   * function mySigner (data) {
   *   return new Promise((resolve, reject) => {
   *     const signature = /// sign it
   *     resolve(signature)
   *   })
   * }
   *
   * const credentials = new Credentials({
   *   did: process.env.APPLICATION_DID,
   *   signer: mySigner
   * })
   *
   * @param       {Object}            [settings]             setttings
   * @param       {DID}               settings.did           Application [DID](https://w3c-ccg.github.io/did-spec/#decentralized-identifiers-dids) (unique identifier) for your application
   * @param       {String}            settings.privateKey    A hex encoded 32 byte private key
   * @param       {SimpleSigner}      settings.signer        a signer object, see [Signer Functions](https://github.com/uport-project/did-jwt#signer-functions)
   * @param       {Object}            settings.ethrConfig    Configuration object for ethr did resolver. See [ethr-did-resolver](https://github.com/uport-project/ethr-did-resolver)
   * @param       {Object}            settings.muportConfig  Configuration object for muport did resolver. See [muport-did-resolver](https://github.com/uport-project/muport-did-resolver)
   * @param       {Address}           settings.address       DEPRECATED your uPort address (may be the address of your application's uPort identity)
   * @param       {Object}            settings.networks      DEPRECATED networks config object, ie. {  '0x94365e3b': { rpcUrl: 'https://private.chain/rpc', address: '0x0101.... }}
   * @param       {UportLite}         settings.registry      DEPRECATED a registry object from UportLite
   * @return      {Credentials}                              self
   */
  constructor ({did, address, privateKey, signer, networks, registry, ethrConfig, muportConfig} = {}) {
    if (signer) {
      this.signer = signer
    } else if (privateKey) {
      this.signer = SimpleSigner(privateKey)

    }
    if (did) {
      this.did = did
    } else if (address) {
      if (MNID.isMNID(address)) {
        this.did = `did:uport:${address}`
      }
      if (address.match('^0x[0-9a-fA-F]{40}$')) {
        this.did = `did:ethr:${address}`
      }
    } else if (privateKey) {
      const kp = secp256k1.keyFromPrivate(privateKey)
      const address = toEthereumAddress(kp.getPublic('hex'))
      this.did = `did:ethr:${address}`
    }

    this.signJWT = (payload, expiresIn) => createJWT(payload, {issuer: this.did, signer: this.signer, alg: this.did.match('^did:uport:') ? 'ES256K' : 'ES256K-R', expiresIn })

    UportDIDResolver(registry || UportLite({networks: networks ? configNetworks(networks) : {}}))
    EthrDIDResolver(ethrConfig || {})
    MuportDIDResolver(muportConfig || {})
  }

  /**
   * generate a DID and private key
   */
  static createIdentity () {
    const kp = secp256k1.genKeyPair()
    const publicKey = kp.getPublic('hex')
    const privateKey = kp.getPrivate('hex')
    const address = toEthereumAddress(publicKey)
    const did = `did:ethr:${address}`
    return {did, privateKey}
  }

/**
 *  Creates a [Selective Disclosure Request JWT](https://github.com/uport-project/specs/blob/develop/messages/sharereq.md)
 *
 *  @example
 *  const req = { requested: ['name', 'country'],
 *                callbackUrl: 'https://myserver.com',
 *                notifications: true }
 *  credentials.requestDisclosure(req).then(jwt => {
 *      ...
 *  })
 *
 *  @param    {Object}             [params={}]           request params object
 *  @param    {Array}              params.requested      an array of attributes for which you are requesting credentials to be shared for
 *  @param    {Array}              params.verified       an array of attributes for which you are requesting verified credentials to be shared for
 *  @param    {Boolean}            params.notifications  boolean if you want to request the ability to send push notifications
 *  @param    {String}             params.callbackUrl    the url which you want to receive the response of this request
 *  @param    {String}             params.network_id     network id of Ethereum chain of identity eg. 0x4 for rinkeby
 *  @param    {String}             params.accountType    Ethereum account type: "general", "segregated", "keypair", "devicekey" or "none"
 *  @param    {Number}             expiresIn             Seconds until expiry
 *  @return   {Promise<Object, Error>}                   a promise which resolves with a signed JSON Web Token or rejects with an error
 */
  requestDisclosure (params = {}, expiresIn = 600) {
    const payload = {}
    if (params.requested) {
      payload.requested = params.requested
    }
    if (params.verified) {
      payload.verified = params.verified
    }
    if (params.notifications) {
      payload.permissions = ['notifications']
    }
    if (params.callbackUrl) {
      payload.callback = params.callbackUrl
    }
    if (params.network_id) {
      payload.net = params.network_id
    }
    if (params.accountType) {
      if (['general', 'segregated', 'keypair', 'devicekey', 'none'].indexOf(params.accountType) >= 0) {
        payload.act = params.accountType
      } else {
        return Promise.reject(new Error(`Unsupported accountType ${params.accountType}`))
      }
    }
    if (params.exp) {
      payload.exp = params.exp
    }
    return this.signJWT({...payload, type: 'shareReq'}, params.exp ? undefined : expiresIn)
  }
/**
 *  Creates a [Selective Disclosure Request JWT](https://github.com/uport-project/specs/blob/develop/messages/sharereq.md)
 *
 *  @example
 *  const req = { requested: ['name', 'country'],
 *                callbackUrl: 'https://myserver.com',
 *                notifications: true }
 *  credentials.createRequest(req).then(jwt => {
 *      ...
 *  })
 *
 *  @param    {Object}             [params={}]           request params object
 *  @param    {Array}              params.requested      an array of attributes for which you are requesting credentials to be shared for
 *  @param    {Array}              params.verified       an array of attributes for which you are requesting verified credentials to be shared for
 *  @param    {Boolean}            params.notifications  boolean if you want to request the ability to send push notifications
 *  @param    {String}             params.callbackUrl    the url which you want to receive the response of this request
 *  @param    {String}             params.network_id     network id of Ethereum chain of identity eg. 0x4 for rinkeby
 *  @param    {String}             params.accountType    Ethereum account type: "general", "segregated", "keypair", "devicekey" or "none"
 *  @return   {Promise<Object, Error>}                   a promise which resolves with a signed JSON Web Token or rejects with an error
 * @deprecated
 */

  createRequest (params = {}) {
    return this.requestDisclosure(params)
  }

/**
 * Creates a [Selective Disclosure Response JWT](https://github.com/uport-project/specs/blob/develop/messages/shareresp.md).
 *
 * This can either be used to share information about the signing identity or as the response to a
 * [Selective Disclosure Flow](https://github.com/uport-project/specs/blob/develop/flows/selectivedisclosure.md), where it can be used to authenticate the identity.
 *
 *  @example
 *  credentials.disclose({own: {name: 'Lourdes Valentina Gomez'}}).then(jwt => {
 *      ...
 *  })
 *
 *  @param    {Object}             [params={}]           request params object
 *  @param    {JWT}                params.req            A selective disclosure Request JWT if this is returned as part of an authentication flow
 *  @param    {Object}             params.own            An object of self attested claims about the signer (eg. name etc)
 *  @param    {Array}              params.verified       An array of attestation JWT's to include
 *  @param    {MNID}               params.nad            An ethereum address encoded as an [MNID](https://github.com/uport-project/mnid)
 *  @param    {Array}              params.capabilities   An array of capability JWT's to include
 *  @return   {Promise<Object, Error>}                   a promise which resolves with a signed JSON Web Token or rejects with an error
 */
async disclose (payload = {}, expiresIn = 600 ) {
  if (payload.req) {
    const verified = await verifyJWT(payload.req)
    if (verified.issuer) {
      payload.aud = verified.issuer
    }
  }
  return this.signJWT({...payload, type: 'shareResp'}, expiresIn)
}

async processDisclosurePayload ({doc, payload}) {
  const credentials = {...doc.uportProfile || {}, ...(payload.own || {}), ...(payload.capabilities && payload.capabilities.length === 1 ? {pushToken: payload.capabilities[0]} : {}), address: payload.iss}
  if (payload.nad) {
    credentials.networkAddress = payload.nad
  }
  if (payload.dad) {
    credentials.deviceKey = payload.dad
  }
  if (payload.verified) {
    const verified = await Promise.all(payload.verified.map(token => verifyJWT(token, {audience: this.did})))
    return {...credentials, verified: verified.map(v => ({...v.payload, jwt: v.jwt}))}
  } else {
    return credentials
  }
}

/**
  *  Authenticates [Selective Disclosure Response JWT](https://github.com/uport-project/specs/blob/develop/messages/shareresp.md) from mobile
  *  app as part of the [Selective Disclosure Flow](https://github.com/uport-project/specs/blob/develop/flows/selectivedisclosure.md).
  *
  *  It Verifies and parses the given response token and verifies the challenge response flow.
  *
  *  @example
  *  const resToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJyZXF1Z....'
  *  credentials.authenticate(resToken).then(res => {
  *      const credentials = res.verified
  *       const name =  res.name
  *      ...
  *  })
  *
  *  @param    {String}                  token                 a response token
  *  @param    {String}                  [callbackUrl=null]    callbackUrl
  *  @return   {Promise<Object, Error>}                        a promise which resolves with a parsed response or rejects with an error.
  */
  async authenticate (token, callbackUrl = null) {
    const { payload, doc } = await verifyJWT(token, {audience: this.did, callbackUrl, auth: true})

    if(payload.req) {
      const challenge = await verifyJWT(payload.req)
      if(challenge.payload.iss === this.did && challenge.payload.type === 'shareReq') {
        return this.processDisclosurePayload({payload, doc})
      }
    } else {
      throw new Error('Challenge was not included in response')
    }
  }

  /**
 *  Creates a signed request for the user to attest a list of claims.
 *
 *  @example
 *  const unsignedClaim = {
 *    claim: {
 *      "Citizen of city X": {
 *        "Allowed to vote": true,
 *        "Document": "QmZZBBKPS2NWc6PMZbUk9zUHCo1SHKzQPPX4ndfwaYzmPW"
 *      }
 *    },
 *    sub: "2oTvBxSGseWFqhstsEHgmCBi762FbcigK5u"
 *  }
 *  credentials.createVerificationRequest(unsignedClaim).then(jwt => {
 *    ...
 *  })
 *
 *  @param    {Object}              unsignedClaim       an object that is an unsigned claim which you want the user to attest
 *  @param    {String}             sub                  the DID of the identity you want to sign the attestation
 *  @return   {Promise<Object, Error>}                  a promise which resolves with a signed JSON Web Token or rejects with an error
 */
  createVerificationRequest(unsignedClaim, sub) {
    return this.signJWT({unsignedClaim, sub, type: 'verReq'})
  }

/**
  *  Receive signed response token from mobile app. Verifies and parses the given response token.
  *
  *  @example
  *  const resToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJyZXF1Z....'
  *  credentials.receive(resToken).then(res => {
  *      const credentials = res.verified
         const name =  res.name
  *      ...
  *  })
  *
  *  @param    {String}                  token                 a response token
  *  @param    {String}                  [callbackUrl=null]    callbackUrl
  *  @return   {Promise<Object, Error>}                        a promise which resolves with a parsed response or rejects with an error.
  *  @deprecated
  */
  receive (token, callbackUrl = null) {
    return this.authenticate(token, callbackUrl)
  }

  /**
  *  Verify and return profile from a [Selective Disclosure Response JWT](https://github.com/uport-project/specs/blob/develop/messages/shareresp.md).
  *
  *  The main difference between this and `authenticate()` is that it does not verify the challenge. This can be used to verify user profiles that have been shared
  *  through other methods such as QR codes and messages.
  *
  *  @example
  *  const resToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJyZXF1Z....'
  *  credentials.verifyProfile(resToken).then(profile => {
  *      const credentials = profile.verified
         const name =  profile.name
  *      ...
  *  })
  *
  *  @param    {String}                  token                 a response token
  *  @return   {Promise<Object, Error>}                        a promise which resolves with a parsed response or rejects with an error.
  */
 async verifyProfile (token) {
  const { payload, doc } = await verifyJWT(token, {audience: this.did})
  return this.processDisclosurePayload({ payload, doc })
}

/**
  *  Create a credential (a signed JSON Web Token)
  *
  *  @example
  *  credentials.attest({
  *   sub: '5A8bRWU3F7j3REx3vkJ...', // uPort address of user, likely a MNID
  *   exp: <future timestamp>,
  *   claim: { name: 'John Smith' }
  *  }).then( credential => {
  *   ...
  *  })
  *
  * @param    {Object}            [credential]           a unsigned credential object
  * @param    {String}            credential.sub         subject of credential (a uPort address)
  * @param    {String}            credential.claim       claim about subject single key value or key mapping to object with multiple values (ie { address: {street: ..., zip: ..., country: ...}})
  * @param    {String}            credential.exp         time at which this claim expires and is no longer valid (seconds since epoch)
  * @return   {Promise<Object, Error>}                   a promise which resolves with a credential (JWT) or rejects with an error
  */
  attest ({sub, claim, exp}) {
    return this.signJWT({sub: sub, claim, exp})
  }

// /**
//   *  Look up a profile in the registry for a given uPort address. Address must be MNID encoded.
//   *
//   *  @example
//   *  credentials.lookup('5A8bRWU3F7j3REx3vkJ...').then(profile => {
//   *     const name = profile.name
//   *     const pubkey = profile.pubkey
//   *     ...
//   *   })
//   *
//   * @param    {String}            address             a MNID encoded address
//   * @return   {Promise<Object, Error>}                a promise which resolves with parsed profile or rejects with an error
//   */
//   lookup (address) {
//     return this.settings.registry(address)
//   }
}

const configNetworks = (nets) => {
  Object.keys(nets).forEach((key) => {
    const net = nets[key]
    if (typeof net === 'object') {
      ['registry', 'rpcUrl'].forEach((key) => {
        if (!net.hasOwnProperty(key)) throw new Error(`Malformed network config object, object must have '${key}' key specified.`)
      })
    } else {
      throw new Error(`Network configuration object required`)
    }
  })
  return nets
}


export default Credentials
