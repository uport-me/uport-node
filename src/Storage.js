import redis from 'redis'

class Storage {

  // options.host options.port
  constructor(options){
    const opts = options || {}
    if (!redis.createClient) throw new Error('Storage: Redis is not available in this enviroment, use functionality which does not require persistent storage')
    this.client = redis.createClient(opts)
    this.client.on("error", err => {console.log("Error " + err)})
    this.challengeKeyPrefix = opts.challengeKeyPrefix || `challenge`
    this.responseKeyPrefix = opts.responseKeyPrefix || `response`
    this.challengeKey = pairId => (`${this.challengeKeyPrefix}:${pairId}`)
    this.responseKey = pairId => (`${this.responseKeyPrefix}:${pairId}`)
  }

  writeChallenge(pairId, challenge) {
  return this.set(this.challengeKey(pairId), challenge)
  }

  readChallenge(pairId) {
    return this.get(this.challengeKey(pairId))
  }

  writeResponse(pairId, response) {
    return this.set(this.responseKey(pairId), response)
  }

  readResponse(pairId) {
    return this.get(this.responseKey(pairId))
  }

  get(key) {
    return new Promise((resolve, reject) => {
      this.client.get(key, (err, res) => {
        if (err) reject(err)
        return resolve(res)
      })
    })
  }

  set(key, value) {
    return new Promise((resolve, reject) => {
      this.client.set(key, value, (err, res) => {
        if (err) reject(err)
        return resolve(res)
      })
    })
  }
}

export default Storage
