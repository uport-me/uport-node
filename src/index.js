import Credentials from './Credentials'
import { createJWT, verifyJWT } from './JWT'
import { SimpleSigner } from 'did-jwt'
import { normalizeDID } from 'did-jwt/lib/JWT'
import { Contract, ContractFactory } from './Contract'
import UportDIDResolver from 'uport-did-resolver'

const JWT = { createJWT: createJWT , verifyJWT: verifyJWT }

module.exports = { Credentials, SimpleSigner, Contract, ContractFactory, JWT, normalizeDID }
