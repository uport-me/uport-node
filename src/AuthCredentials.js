import Credentials from './Credentials.js'
import Storage from './Storage.js'
import crypto from 'crypto'
import { createJWT } from './JWT'

// Generate random 128 byte hex string
const random  = () => {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(32, (err, buf) => {
      if (err) reject(error)
      resolve(buf.toString('hex'))
    })
  })
}

class AuthCredentials extends Credentials {
  constructor (settings = {}) {
    super(settings)
    this.storage = settings.storage || new Storage()
    this.random = settings.random || random
  }

  // Create request token
  createRequest (params = {}) {
    const payload = {}
    if (params.requested) {
      payload.requested = params.requested
    }
    if (params.verified) {
      payload.verified = params.verified
    }
    if (params.notifications) {
      payload.permissions = ['notifications']
    }
    if (params.callbackUrl) {
      payload.callback = params.callbackUrl
    }

    return Promise.all([this.random(), this.random()]).then(rand => {
      payload.challenge = rand[0]
      payload.pairId = rand[1]
    }).then(() => this.storage.writeChallenge(payload.pairId, payload.challenge)
    ).then(res => createJWT(this.settings, {...payload, type: 'shareReq'}))
  }

  receive (token, response = '', callbackUrl = null) {
    // Verify sig and check challenge equivalence
    let credentials
    return super.receive(token, callbackUrl).then(res => {
      credentials = res
      return this.storage.readChallenge(credentials.pairId)
    }).then(val => {
      const challenge = val
      if (challenge === credentials.challenge) {
        return this.storage.writeResponse(credentials.pairId, response)
      }
      this.storage.writeResponse(credentials.pairId, 'Error')
      throw new Error('Authentication Failed: receive() Challenge did not match')
    }).then(res => credentials)
  }

  authResponse(pairId) {
    return this.storage.readResponse(pairId)
  }
}

export default AuthCredentials
