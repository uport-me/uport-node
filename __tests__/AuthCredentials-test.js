import AuthCredentials from '../src/AuthCredentials'
import { createJWT } from '../src/JWT'
import SimpleSigner from '../src/SimpleSigner'
import { SECP256K1Client, TokenVerifier, decodeToken } from 'jsontokens'
const sinon = require('sinon')
import nock from 'nock'
import redisMock from 'redis-mock'
import MockDate from 'mockdate'
MockDate.set(1485321133996)

const privateKey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const publicKey = SECP256K1Client.privateKeyToPublicKey(privateKey)
const signer = SimpleSigner(privateKey)
const verifier = new TokenVerifier('ES256K', publicKey)
const profileA = {publicKey, name: 'David Chaum'}
const registry = (address) => new Promise((resolve, reject) => resolve(address === '0x001122' ? profileA : null))
const uport = new AuthCredentials({signer, address: '0x001122', registry})
const rand1 = 'db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8'
const rand2 = '6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac'
const challenge = rand1
const pairId = rand2
const invalidChallenge = rand1
const challengeKeyPrefix = `challenge`

describe('createRequest', () => {

  let uport, redis

  beforeEach(() => {
    uport = new AuthCredentials({signer, address: '0x001122', registry})

    const random = sinon.stub()
    random.onCall(0).returns(rand1);
    random.onCall(1).returns(rand2);

    uport.random = random
    uport.storage.client = redisMock.createClient()
    redis = uport.storage.client
  })

  it('creates a valid JWT for a request', () => {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return expect(verifier.verify(jwt)).toBeTruthy()
    })
  })

  it('has correct payload in JWT for a plain request for public details', () => {
    return uport.createRequest().then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('ignores unsupported request parameters', () => {
    return uport.createRequest({signing: true, sellSoul: true}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request', () => {
    return uport.createRequest({requested: ['name', 'phone']}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request asking for verified credentials', () => {
    return uport.createRequest({requested: ['name', 'phone'], verified: ['name']}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request with callbackUrl', () => {
    return uport.createRequest({ requested: ['name', 'phone'], callbackUrl: 'https://myserver.com'}).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('has correct payload in JWT for a request for push notifications', () => {
    return uport.createRequest({ requested: ['name', 'phone'], notifications: true }).then((jwt) => {
      return expect(decodeToken(jwt)).toMatchSnapshot()
    })
  })

  it('inlcudes a challenge and pairId in the payload in the JWT', () => {
    return uport.createRequest({ requested: ['name', 'phone']}).then((jwt) => {
      const decodedJWT = decodeToken(jwt)
      expect(decodedJWT.payload.challenge).toEqual(challenge)
      expect(decodedJWT.payload.pairId).toEqual(pairId)
    })
  })

  it('stores a challenge with the pairId in storage', (done) => {
    uport.createRequest({ requested: ['name', 'phone']}).then((jwt) => {
      redis.get(`${challengeKeyPrefix}:${pairId}`, (err, res) => {
          if (err) throw new Error('Redis client could not get value')
          expect(res).toEqual(challenge)
          done()
      })
    })
  })
})

describe('receive', () => {

  let uport, redis

  beforeAll((done) => {
    uport = new AuthCredentials({signer, address: '0x001122', registry})
    uport.storage.client = redisMock.createClient()
    redis = uport.storage.client

    redis.set(`${challengeKeyPrefix}:${pairId}`, challenge, (err, res) => {
        if (err) throw new Error('Redis client could not set value')
        done()
    })
  })

  function createShareResp (payload = {}) {
    return createJWT({address: '0x001122', signer}, {...payload, challenge, pairId, type: 'shareResp'})
  }

  function createInvalidShareResp (payload = {}) {
    return createJWT({address: '0x001122', signer}, {...payload, challenge: '123456789101121314151617181920', pairId, type: 'shareResp'})
  }

  function createShareRespWithVerifiedCredential (payload = {}, verifiedClaim = {sub: '0x112233', claim: {email: 'bingbangbung@email.com'}, exp: 1485321133996 + 1000}) {
    return uport.attest(verifiedClaim).then(jwt => {
      return createShareResp({...payload, challenge, pairId, verified: [jwt]})
    })
  }

  // All tests only return if the challenge is valid, otherwise they would fail

  it('returns profile mixing public and private claims', () => {
    return createShareResp({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile mixing public and private claims and verified credentials', () => {
    return createShareRespWithVerifiedCredential({own: {name: 'Davie', phone: '+15555551234'}}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns profile with only public claims', () => {
    return createShareResp().then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile).toMatchSnapshot()
    )
  })

  it('returns pushToken if available', () => {
    return createShareResp({capabilities: ['PUSHTOKEN']}).then(jwt => uport.receive(jwt)).then(profile =>
      expect(profile.pushToken).toEqual('PUSHTOKEN')
    )
  })

  it('throws an error if the challenge is not valid', () => {
    return createInvalidShareResp().then(jwt => uport.receive(jwt)).then(profile =>
      fail()
    ).catch(err => {
      expect(err).toEqual(expect.stringMatching('Authentication Failed'))
    })
  })
})
