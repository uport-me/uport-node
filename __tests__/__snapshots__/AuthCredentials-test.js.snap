exports[`createRequest has correct payload in JWT for a plain request for public details 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "type": "shareReq",
  },
  "signature": "IeA-zYbNqqnNiVy2SsmwLIvduWhdtTmRasEElzazg9tYSyOujGEnn6TOtPBVTR9WqgNhUyIqiJIlTqIa3aE_mA",
}
`;

exports[`createRequest has correct payload in JWT for a request 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "rybln1wqA9LzcHynGQVZxXzBRpKGNuzOaDtf64G1HVQQ7QNRXjaL4599HZs_qUJYs4euRYIN5wj2xHdaXZZLWg",
}
`;

exports[`createRequest has correct payload in JWT for a request asking for verified credentials 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
    "verified": Array [
      "name",
    ],
  },
  "signature": "1VD8zUL1gXKoNrpaJNVEQgd90TT_AuOznk9qihAUizUxoyh4GbnOldUsBwwlkL1tKTVaAWFqxZTHiqZ1NuZ5mA",
}
`;

exports[`createRequest has correct payload in JWT for a request for push notifications 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "permissions": Array [
      "notifications",
    ],
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "6NMMMDaL3CjyfS9ciTKOvmVNnS4m6rjDXNM2ZJE8CQqqIbgGooRwrOVTbUt17LqxbbgSLQSdff0LL_FKbyoW4Q",
}
`;

exports[`createRequest has correct payload in JWT for a request with callbackUrl 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "callback": "https://myserver.com",
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "requested": Array [
      "name",
      "phone",
    ],
    "type": "shareReq",
  },
  "signature": "UPBxIl7JsDWRU_mG5jf7bX7OmKuUBiy6OD5E9lFmz1V2rXcYh9AAeZqd8SttAL-QNg-kUp7JiODiH9K-br_j6w",
}
`;

exports[`createRequest ignores unsupported request parameters 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
    "iat": 1485321133996,
    "iss": "0x001122",
    "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
    "type": "shareReq",
  },
  "signature": "IeA-zYbNqqnNiVy2SsmwLIvduWhdtTmRasEElzazg9tYSyOujGEnn6TOtPBVTR9WqgNhUyIqiJIlTqIa3aE_mA",
}
`;

exports[`receive returns profile mixing public and private claims 1`] = `
Object {
  "address": "0x001122",
  "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
  "name": "Davie",
  "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;

exports[`receive returns profile mixing public and private claims and verified credentials 1`] = `
Object {
  "address": "0x001122",
  "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
  "name": "Davie",
  "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
  "phone": "+15555551234",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
  "verified": Array [
    Object {
      "claim": Object {
        "email": "bingbangbung@email.com",
      },
      "exp": 1485321134996,
      "iat": 1485321133996,
      "iss": "0x001122",
      "jwt": "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJzdWIiOiIweDExMjIzMyIsImNsYWltIjp7ImVtYWlsIjoiYmluZ2JhbmdidW5nQGVtYWlsLmNvbSJ9LCJleHAiOjE0ODUzMjExMzQ5OTYsImlzcyI6IjB4MDAxMTIyIiwiaWF0IjoxNDg1MzIxMTMzOTk2fQ.-mEzVMPYnzqFhOr0O7fs71-dWAacnllVyOdWQY0zh2ZdIt7-30IYTewds4tGlkLmMky-Y1ZjRmIsxmM7xvAgxg",
      "sub": "0x112233",
    },
  ],
}
`;

exports[`receive returns profile with only public claims 1`] = `
Object {
  "address": "0x001122",
  "challenge": "db2ad7a398dfdd897f5b7ebac6e5995d482a824917c8251cfe30c86926510fa8",
  "name": "David Chaum",
  "pairId": "6d98e4827b8912cf4b061d0d16ac68577b6e0632ff2a4ce89ed70c537334f0ac",
  "publicKey": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
}
`;
